File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/TVA.Core/Scheduling/Schedule.vb
Imports System.Text
Imports TVA.Text.Common
Imports TVA.DateTime

Namespace Scheduling

    Public Class Schedule

#Region " Member declaration "

        Private m_name As String
        Private m_description As String
        Private m_minutePart As SchedulePart
        Private m_hourPart As SchedulePart
        Private m_dayPart As SchedulePart
        Private m_monthPart As SchedulePart
        Private m_dayOfWeekPart As SchedulePart
        Private m_lastDueAt As System.DateTime

#End Region

#Region " Code Scope: Public "

        Public Sub New(ByVal name As String)

            MyClass.New(name, "* * * * *")

        End Sub

        Public Sub New(ByVal name As String, ByVal rule As String)

            MyClass.New(name, rule, "")

        End Sub

        Public Sub New(ByVal name As String, ByVal rule As String, ByVal description As String)

            MyBase.New()
            Me.Name = name
            Me.Rule = rule
            Me.Description = description

        End Sub

        ''' <summary>
        ''' Gets or sets the schedule name.
        ''' </summary>
        ''' <value></value>
        ''' <returns>The schedule name.</returns>
        Public Property Name() As String
            Get
                Return m_name
            End Get
            Set(ByVal value As String)
                If Not String.IsNullOrEmpty(value) Then
                    m_name = value
                Else
                    Throw New ArgumentNullException("Name")
                End If
            End Set
        End Property

        ''' <summary>
        ''' Gets or sets the schedule rule.
        ''' </summary>
        ''' <value></value>
        ''' <returns>The schedule rule.</returns>
        Public Property Rule() As String
            Get
                Return m_minutePart.Text & " " & m_hourPart.Text & " " & m_dayPart.Text & " " & m_monthPart.Text & " " & m_dayOfWeekPart.Text
            End Get
            Set(ByVal value As String)
                If Not String.IsNullOrEmpty(value) Then
                    With RemoveDuplicateWhiteSpace(value).Split(" "c)
                        If .Length = 5 Then
                            m_minutePart = New SchedulePart(.GetValue(0).ToString(), DateTimePart.Minute)
                            m_hourPart = New SchedulePart(.GetValue(1).ToString(), DateTimePart.Hour)
                            m_dayPart = New SchedulePart(.GetValue(2).ToString(), DateTimePart.Day)
                            m_monthPart = New SchedulePart(.GetValue(3).ToString(), DateTimePart.Month)
                            m_dayOfWeekPart = New SchedulePart(.GetValue(4).ToString(), DateTimePart.DayOfWeek)

                            ' Update the schedule description.
                            With New StringBuilder()
                                .Append(m_minutePart.Description)
                                .Append(", ")
                                .Append(m_hourPart.Description)
                                .Append(", ")
                                .Append(m_dayPart.Description)
                                .Append(", ")
                                .Append(m_monthPart.Description)
                                .Append(", ")
                                .Append(m_dayOfWeekPart.Description)

                                m_description = .ToString()
                            End With
                        Else
                            Throw New ArgumentException("Schedule rule must have exactly 5 parts (Example: * * * * *).")
                        End If
                    End With
                Else
                    Throw New ArgumentNullException("Rule")
                End If
            End Set
        End Property

        Public Property Description() As String
            Get
                Return m_description
            End Get
            Set(ByVal value As String)
                If Not String.IsNullOrEmpty(value) Then
                    m_description = value
                End If
            End Set
        End Property

        Public ReadOnly Property LastDueAt() As System.DateTime
            Get
                Return m_lastDueAt
            End Get
        End Property

        Public ReadOnly Property MinutePart() As SchedulePart
            Get
                Return m_minutePart
            End Get
        End Property

        Public ReadOnly Property HourPart() As SchedulePart
            Get
                Return m_hourPart
            End Get
        End Property

        Public ReadOnly Property DayPart() As SchedulePart
            Get
                Return m_dayPart
            End Get
        End Property

        Public ReadOnly Property MonthPart() As SchedulePart
            Get
                Return m_monthPart
            End Get
        End Property

        Public ReadOnly Property DaysOfWeekPart() As SchedulePart
            Get
                Return m_dayOfWeekPart
            End Get
        End Property

        ''' <summary>
        ''' Gets the current status of the schedule.
        ''' </summary>
        ''' <value></value>
        ''' <returns>The current status of the schedule.</returns>
        Public ReadOnly Property Status() As String
            Get
                With New System.Text.StringBuilder()
                    .Append("             Schedule name: ")
                    .Append(m_name)
                    .AppendLine()
                    .Append("             Schedule rule: ")
                    .Append(Rule)
                    .AppendLine()
                    .Append("             Last run time: ")
                    .Append(IIf(m_lastDueAt = System.DateTime.MinValue, "Never", m_lastDueAt))
                    .AppendLine()

                    Return .ToString()
                End With
            End Get
        End Property

        ''' <summary>
        ''' Checks whether the schedule is due at the present system time.
        ''' </summary>
        ''' <returns>True if the schedule is due at the present system time; otherwise False.</returns>
        Public Function IsDue() As Boolean

            Dim currentDateTime As System.DateTime = System.DateTime.Now
            If m_minutePart.Matches(currentDateTime) AndAlso m_hourPart.Matches(currentDateTime) AndAlso _
                    m_dayPart.Matches(currentDateTime) AndAlso m_monthPart.Matches(currentDateTime) AndAlso _
                    m_dayOfWeekPart.Matches(currentDateTime) Then
                m_lastDueAt = currentDateTime
                Return True
            Else
                Return False
            End If

        End Function

        Public Overrides Function Equals(ByVal obj As Object) As Boolean

            Dim other As Schedule = TryCast(obj, Schedule)
            If other IsNot Nothing AndAlso other.Name = Me.Name AndAlso other.Rule = Me.Rule Then
                Return True
            Else
                Return False
            End If

        End Function

#End Region

    End Class

End Namespace

File: Source/OldCode/Archive/2005 TVA Code Library/Version/CS/TVA.Core_CS/Scheduling/Schedule.cs
using System.Diagnostics;
using System.Linq;
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.Text;
//using TVA.Text.Common;
using TVA.DateTime;


namespace TVA
{
	namespace Scheduling
	{
		
		public class Schedule
		{
			
			
			#region " Member declaration "
			
			private string m_name;
			private string m_description;
			private SchedulePart m_minutePart;
			private SchedulePart m_hourPart;
			private SchedulePart m_dayPart;
			private SchedulePart m_monthPart;
			private SchedulePart m_dayOfWeekPart;
			private System.DateTime m_lastDueAt;
			
			#endregion
			
			#region " Code Scope: Public "
			
			public Schedule(string name) : this(name, "* * * * *")
			{
				
				
			}
			
			public Schedule(string name, string rule) : this(name, rule, "")
			{
				
				
			}
			
			public Schedule(string name, string rule, string description)
			{
				
				this.Name = name;
				this.Rule = rule;
				this.Description = description;
				
			}
			
			/// <summary>
			/// Gets or sets the schedule name.
			/// </summary>
			/// <value></value>
			/// <returns>The schedule name.</returns>
			public string Name
			{
				get
				{
					return m_name;
				}
				set
				{
					if (! string.IsNullOrEmpty(value))
					{
						m_name = value;
					}
					else
					{
						throw (new ArgumentNullException("Name"));
					}
				}
			}
			
			/// <summary>
			/// Gets or sets the schedule rule.
			/// </summary>
			/// <value></value>
			/// <returns>The schedule rule.</returns>
			public string Rule
			{
				get
				{
					return m_minutePart.Text + " " + m_hourPart.Text + " " + m_dayPart.Text + " " + m_monthPart.Text + " " + m_dayOfWeekPart.Text;
				}
				set
				{
					if (! string.IsNullOrEmpty(value))
					{
						string with_1 = TVA.Text.Common.RemoveDuplicateWhiteSpace(value).Split(' ');
						if (with_1.Length == 5)
						{
							m_minutePart = new SchedulePart(with_1.GetValue(0).ToString(), DateTimePart.Minute);
							m_hourPart = new SchedulePart(with_1.GetValue(1).ToString(), DateTimePart.Hour);
							m_dayPart = new SchedulePart(with_1.GetValue(2).ToString(), DateTimePart.Day);
							m_monthPart = new SchedulePart(with_1.GetValue(3).ToString(), DateTimePart.Month);
							m_dayOfWeekPart = new SchedulePart(with_1.GetValue(4).ToString(), DateTimePart.DayOfWeek);
							
							// Update the schedule description.
							System.Text.StringBuilder with_2 = new StringBuilder();
							with_2.Append(m_minutePart.Description);
							with_2.Append(", ");
							with_2.Append(m_hourPart.Description);
							with_2.Append(", ");
							with_2.Append(m_dayPart.Description);
							with_2.Append(", ");
							with_2.Append(m_monthPart.Description);
							with_2.Append(", ");
							with_2.Append(m_dayOfWeekPart.Description);
							
							m_description = with_2.ToString();
						}
						else
						{
							throw (new ArgumentException("Schedule rule must have exactly 5 parts (Example: * * * * *)."));
						}
					}
					else
					{
						throw (new ArgumentNullException("Rule"));
					}
				}
			}
			
			public string Description
			{
				get
				{
					return m_description;
				}
				set
				{
					if (! string.IsNullOrEmpty(value))
					{
						m_description = value;
					}
				}
			}
			
			public System.DateTime LastDueAt
			{
				get
				{
					return m_lastDueAt;
				}
			}
			
			public SchedulePart MinutePart
			{
				get
				{
					return m_minutePart;
				}
			}
			
			public SchedulePart HourPart
			{
				get
				{
					return m_hourPart;
				}
			}
			
			public SchedulePart DayPart
			{
				get
				{
					return m_dayPart;
				}
			}
			
			public SchedulePart MonthPart
			{
				get
				{
					return m_monthPart;
				}
			}
			
			public SchedulePart DaysOfWeekPart
			{
				get
				{
					return m_dayOfWeekPart;
				}
			}
			
			/// <summary>
			/// Gets the current status of the schedule.
			/// </summary>
			/// <value></value>
			/// <returns>The current status of the schedule.</returns>
			public string Status
			{
				get
				{
					System.Text.StringBuilder with_1 = new System.Text.StringBuilder();
					with_1.Append("             Schedule name: ");
					with_1.Append(m_name);
					with_1.AppendLine();
					with_1.Append("             Schedule rule: ");
					with_1.Append(Rule);
					with_1.AppendLine();
					with_1.Append("             Last run time: ");
					with_1.Append(m_lastDueAt == System.DateTime.MinValue ? "Never" : m_lastDueAt);
					with_1.AppendLine();
					
					return with_1.ToString();
				}
			}
			
			/// <summary>
			/// Checks whether the schedule is due at the present system time.
			/// </summary>
			/// <returns>True if the schedule is due at the present system time; otherwise False.</returns>
			public bool IsDue()
			{
				
				System.DateTime currentDateTime = System.DateTime.Now;
				if (m_minutePart.Matches(currentDateTime) && m_hourPart.Matches(currentDateTime) && m_dayPart.Matches(currentDateTime) && m_monthPart.Matches(currentDateTime) && m_dayOfWeekPart.Matches(currentDateTime))
				{
					m_lastDueAt = currentDateTime;
					return true;
				}
				else
				{
					return false;
				}
				
			}
			
			public override bool Equals(object obj)
			{
				
				Schedule other = obj as Schedule;
				if ((other != null)&& other.Name == this.Name && other.Rule == this.Rule)
				{
					return true;
				}
				else
				{
					return false;
				}
				
			}
			
			#endregion
			
		}
		
	}
}
